source/xcb.c

```diff
@@ -1375,6 +1404,15 @@ static gboolean main_loop_x11_event_handler(xcb_generic_event_t *ev,
   if (xcb->sndisplay != NULL) {
     sn_xcb_display_process_event(xcb->sndisplay, ev);
   }
+
+  if (xcb->im && xcb_xim_filter_event(xcb->im, ev) == 0) {
+    if (xcb->ic &&
+        (type == XCB_KEY_PRESS || type == XCB_KEY_RELEASE || type == 31)) {
+      xcb_xim_forward_event(xcb->im, xcb->ic, (xcb_key_press_event_t *)ev);
```

Comment:
Is it safe to pass `KEY_RELEASE` events here? (the cast is mostly safe because the structs are built that way, but maybe we shouldn’t still)

Reply:
not sure, but the xcb_xim_forward_event only accepts xcb_key_press_event_t, and both implementations I'm using as reference: 
https://github.com/H-M-H/xcb-imdkit-rs/blob/master/src/lib.rs
https://github.com/fcitx/xcb-imdkit/blob/30e2f16f9a8b0e338e25ce5e3643809a07ad41f0/test/client_demo.c

just pass both KEY_PRESS and KEY_RELEASE